### PR TITLE
feat: Track source state origins more precisely

### LINKED ISSUE #2177: adding an external, and then doing chezmoi forget "that external", asks to remove all chezmoi
## Describe the bug

Adding an external, for example, a github repo into "test_external" in chezmoi.
Then, doing chezmoi forget test_external, chezmoi asks to remove all of chezmoi data.

I tested answering yes, and it did removed everything (no worries though, i did a backup first)

## To reproduce

add an external to .chezmoiexternal.toml.
either apply that external or not, its indiferent.
do a chezmoi forget "that external"

## Expected behavior

chezmoi to either remove the external, from the .chezmoiexternal.toml or simply say that it can't forget it because its a configured external in the file .chezmoiexternal.toml

## Output of command with the `--verbose` flag
(edit: forgot to mention, but I aliased chezmoi to cm, so cm command below is chezmoi)
(using oh-my-posh, nothing against it, nor do I want to forget it :) just an example)
```console
cm --verbose forget oh-my-posh
Remove /mnt/data/alex/chezmoi [yes,no,all,quit]? 
```

## Output of `chezmoi doctor`
in this case, yes, I have the .config/chezmoi.toml added, but I just read the issue where it should not be added, so I will remove it :)
<details>

```console
chezmoi doctor
RESULT    CHECK                MESSAGE
warning   version              v2.18.1, built at 2022-06-23T21:17:54Z
ok        latest-version       v2.18.1
ok        os-arch              linux/amd64 (Garuda Linux)
ok        uname                Linux stormtrooper 5.18.6-zen1-1.1-zen #1 ZEN SMP PREEMPT_DYNAMIC Thu, 23 Jun 2022 15:04:31 +0000 x86_64 GNU/Linux
ok        go-version           go1.18.3 (gc)
ok        executable           /usr/bin/chezmoi
ok        config-file          ~/.config/chezmoi/chezmoi.toml, last modified 2022-06-13T04:51:52+01:00
ok        source-dir           /mnt/data/alex/chezmoi is a git working tree (clean)
warning   suspicious-entries   /mnt/data/alex/chezmoi/dot_config/chezmoi/chezmoi.toml
ok        working-tree         /mnt/data/alex/chezmoi is a git working tree (clean)
ok        dest-dir             ~ is a directory
ok        shell-command        found /bin/bash
ok        shell-args           /bin/bash
ok        cd-command           found /usr/bin/fish
ok        cd-args              /usr/bin/fish
ok        edit-command         found /usr/bin/nvim
ok        edit-args            nvim
info      diff-command         not set
ok        umask                022
ok        git-command          found /usr/bin/git, version 2.36.1
ok        merge-command        found /usr/bin/nvim
info      age-command          age not found in $PATH
ok        gpg-command          found /usr/bin/gpg, version 2.2.35
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      gopass-command       gopass not found in $PATH
ok        keepassxc-command    found /usr/bin/keepassxc-cli, version 2.7.1
info      keeper-command       keeper not found in $PATH
info      keepassxc-db         not set
info      lastpass-command     lpass not found in $PATH
ok        pass-command         found /usr/bin/pass, version 1.7.4
info      vault-command        vault not found in $PATH
info      secret-command       not set

```

</details>

## Additional context

Thanks for chezmoi, very nice and cool tool !!!

### PR DESCRIPTION
**Pull request by twpayne**
Fixes #2177 more correctly, as identified by @bradenhilton in https://github.com/twpayne/chezmoi/pull/2179#issuecomment-1167843293.

### CODE REVIEW

**Comment by bradenhilton** [comment]
I'm using the debugger in VS Code with the following config as per https://github.com/twpayne/chezmoi/pull/2179#issuecomment-1167843293:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Launch",
      "type": "go",
      "request": "launch",
      "mode": "debug",
      "program": "${workspaceFolder}",
      "args": [
        "forget",
        "<homeDir>\\AppData\\Roaming\\mpv\\scripts\\sub-select.lua"
      ],
      "console": "integratedTerminal"
    }
  ]
}
```

This is the output I get:

```console
chezmoi: warning: AppData/Roaming/mpv/scripts/sub-select.lua: cannot forget entry from &{file %!s(bool=false) %!s(bool=false) %!s(bool=false) {[]} [] { []}  
[] {[]} %!s(chezmoi.Duration=0) %!s(int=0) https://raw.githubusercontent.com/CogentRedTester/mpv-sub-select/b617cf9183e92619977d0512500ed5fd3a52b2da/sub-select.lua {<homeDir>/.local/share/chezmoi/home/AppData/Roaming/mpv/scripts/.chezmoiexternal.toml}}
```

I also get the same output when running

```console
❯ go build
❯ go run . forget '<homeDir>\AppData\Roaming\mpv\scripts\sub-select.lua'
```

(I need the full path because my git clones are on my D: drive)

**Comment by bradenhilton** [comment]
It looks as expected now:

```console
chezmoi: warning: AppData/Roaming/mpv/scripts/sub-select.lua: cannot forget entry from https://raw.githubusercontent.com/CogentRedTester/mpv-sub-select/b617cf9183e92619977d0512500ed5fd3a52b2da/sub-select.lua via ~/.local/share/chezmoi/home/AppData/Roaming/mpv/scripts/.chezmoiexternal.toml
```

A few questions:

- Long URLs can take up a lot of real estate, do you think hiding the URL behind the verbose or debug flag would be better? I think the fact that an entry is an external is much more relevant than its URL in general for this use case.
- The current message doesn't really give a reason why the entry can't be forgotten, do you think this should be added?

Something like:

```console
chezmoi: warning: <path/to/external>: cannot forget entry of type external
- or -
chezmoi: warning: <path/to/external>: cannot forget entry. reason: entry is of type external
```

Basically, I think informing the user that...

1. The entry is an external.
2. chezmoi cannot forget external entries.

...should be sufficient most of the time, no?

**Comment by twpayne** [comment]
Thanks for the feedback! I've updated the error message based on it. I think it's OK that the error message is quite long because this is likely a very rarely used command and the extra information will hopefully help users understand the problem when they encounter it.
